### PR TITLE
Per-request timeouts.

### PR DESCRIPTION
**Pull request by gianm**

### CODE REVIEW

File: src/test/java/com/metamx/http/client/JankyServersTest.java [hunk 1]
```diff
@@ -168,6 +168,35 @@ public void testHttpSilentServer() throws Throwable
   }
 
   @Test
+  public void testHttpSilentServerWithRequestTimeout() throws Throwable
+  {
+    final Lifecycle lifecycle = new Lifecycle();
+    try {
+      final HttpClientConfig config = HttpClientConfig.builder().withReadTimeout(new Duration(86400L * 365)).build();
```
Comment (drcrallen) [inline src/test/java/com/metamx/http/client/JankyServersTest.java]: ♪ 5 hundred twenty five thousand six hundred minutes ♫

**Comment by drcrallen** [comment]
:+1:

**Comment by fjy** [comment]
@drcrallen u have to merge this, we can't :P

**Comment by drcrallen** [comment]
@fjy I'm waiting to see if @nishantmonu51 or @xvrl want to review

**Comment by xvrl** [comment]
@gianm how about making this a property of the request instead of adding a separate argument?

**Comment by gianm** [comment]
Timeout seemed like more of a "go" thing than a "request" thing to me, but I could move it. anyone else have thoughts?

**Comment by drcrallen** [comment]
I did some cursory digging and couldn't find examples of where timeout is encapsulated in a request that was intended to be enforced in the client (as opposed to on the server side).

There are lots of places where timeout is a property of the client or connection when intended to be enforced on the client side. As such I think it makes sense to keep with general industry convention and put it as a property of the connection. It might make sense to have a "connection"-like interface at some point, but our "go" effectively works as a connection constructor where our connection object is simply a future. As such I think a parameter to "go" is the best substitute for a connection that we have and is appropriate to have the options for client-side timeout.

**Comment by drcrallen** [comment]
On a side note, it might make sense to make the future have a ListenableFuture callback that handles if the future is cancelled and the connection needs cleaned up (effectively a `connection.close()`) But that's outside the scope of this PR

**Pull request merged**